En este archivo puedes escribir lo que estimes conveniente. Te recomendamos detallar tu solución y todas las suposiciones que estás considerando. Aquí puedes ejecutar las funciones que definiste en los otros archivos de la carpeta src, medir el tiempo, memoria, etc.

In [1]:
import os
import io
import cProfile
import pstats
import q1_time as q1t
import q1_memory as q1m
import q2_time as q2t
import q2_memory as q2m
import q3_time as q3t
import q3_memory as q3m
import json

file_path = '../farmers-protest-tweets-2021-2-4.json'



**En este desafio se busca optimizar tanto tiempos de ejecución como memoria utilizada para 3 preguntas, en particular, dentro de este Notebook evaluaremos los tiempos de ejecución y memoria utilizada**

In [2]:
#Cargar la extension de memory_profiler
%load_ext memory_profiler

In [3]:
#Preparar perfilador de python profiler
pr = cProfile.Profile()

In [4]:
# q1t
pr.enable()

resultq1t = q1t.q1_time(file_path)

pr.disable()

# Primero: se crea un objeto StringIO para capturar la salida del análisis de rendimiento
# Segundo: se crea un objeto Stats y usa print_stats para imprimir el informe
s = io.StringIO()
ps = pstats.Stats(pr, stream=s).sort_stats('cumulative')
ps.print_stats()

your 131072x1 screen size is bogus. expect trouble
24/03/23 19:02:53 WARN Utils: Your hostname, DESKTOP-RA7LUUP resolves to a loopback address: 127.0.1.1; using 172.25.159.50 instead (on interface eth0)
24/03/23 19:02:53 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/03/23 19:02:55 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [5]:

print(s.getvalue())

         18395 function calls (18215 primitive calls) in 12.142 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000   12.142    6.071 /home/josed/Desafio/venv/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3541(run_code)
        2    0.000    0.000   12.142    6.071 {built-in method builtins.exec}
        1    0.000    0.000   12.142   12.142 /tmp/ipykernel_152291/1963347963.py:1(<module>)
        1    0.000    0.000   12.142   12.142 /home/josed/Desafio/src/q1_time.py:8(q1_time)
      217    0.001    0.000    9.424    0.043 /usr/lib/python3.10/socket.py:691(readinto)
      217    9.423    0.043    9.423    0.043 {method 'recv_into' of '_socket.socket' objects}
      212    0.001    0.000    9.416    0.044 /home/josed/Desafio/venv/lib/python3.10/site-packages/py4j/java_gateway.py:1015(send_command)
      213    0.002    0.000    9.414    0.044 /home/josed/Desafio/venv/lib/python3.10/

In [6]:
resultq1t


[(datetime.date(2021, 2, 19), 'Preetm91'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 21), 'Surrypuria'),
 (datetime.date(2021, 2, 18), 'rebelpacifist'),
 (datetime.date(2021, 2, 19), 'KaurDosanjh1979'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 15), 'jot__b')]

En el output de los tiempos de ejecucion de la funcion q1_time se puede observar que la información de la función ejecutada aparece en la linea 4, para el .

In [7]:
%%memit

resultq1m = q1m.q1_memory(file_path)


peak memory: 180.89 MiB, increment: 69.38 MiB


In [8]:
#q1m
resultq1m

[(datetime.date(2021, 2, 19), 'Preetm91'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 21), 'Surrypuria'),
 (datetime.date(2021, 2, 18), 'rebelpacifist'),
 (datetime.date(2021, 2, 19), 'KaurDosanjh1979'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 15), 'jot__b')]


 **Con "%%memit" podemos medir la memoria utilizada antes y despues de ejecutar la funcion**

Utilizamos memory profiler para medir el uso de nuestra memoria, observamos que el pico de memoria fue de 182.25MiB y la memoria adicional utilizada por la ejecución de la función fue de 69.4 MiB, dicho incremento indica que cuanta memoria adicional se utilizo en comparación al uso de memoria antes de ejecutar el codigo.

In [9]:
# q2t
pr.enable()

resultq2t=q2t.q2_time(file_path)

pr.disable()

s2 = io.StringIO()

ps2 = pstats.Stats(pr, stream=s2).sort_stats('cumulative')
ps2.print_stats()

In [10]:
print(s2.getvalue())

         1947130 function calls (1946596 primitive calls) in 18.157 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        4    0.000    0.000   18.157    4.539 /home/josed/Desafio/venv/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3541(run_code)
        4    0.000    0.000   18.157    4.539 {built-in method builtins.exec}
        1    0.000    0.000   12.142   12.142 /tmp/ipykernel_152291/1963347963.py:1(<module>)
        1    0.000    0.000   12.142   12.142 /home/josed/Desafio/src/q1_time.py:8(q1_time)
      217    0.001    0.000    9.424    0.043 /usr/lib/python3.10/socket.py:691(readinto)
      217    9.423    0.043    9.423    0.043 {method 'recv_into' of '_socket.socket' objects}
      212    0.001    0.000    9.416    0.044 /home/josed/Desafio/venv/lib/python3.10/site-packages/py4j/java_gateway.py:1015(send_command)
      213    0.002    0.000    9.414    0.044 /home/josed/Desafio/venv/lib/python3

In [11]:
resultq2t

[('🙏', 2001),
 ('❤', 1402),
 ('🌾', 545),
 ('💚', 498),
 ('😂', 488),
 ('👍', 461),
 ('👉', 450),
 ('✊', 436),
 ('🙏🙏', 406),
 ('👇', 375)]

In [12]:
%%memit
resultq2m = q2m.q2_memory(file_path)


peak memory: 876.09 MiB, increment: 0.10 MiB


In [13]:
#q2m
resultq2m

[('🙏', 2001),
 ('❤', 1402),
 ('🌾', 545),
 ('💚', 498),
 ('😂', 488),
 ('👍', 461),
 ('👉', 450),
 ('✊', 436),
 ('🙏🙏', 406),
 ('👇', 375)]

En este caso se puede observar que el pico de memoria fue de 167.7MiB y la memoria adicional utilizada por la ejecución de la función fue de 8.0 MiB, dicho incremento indica que cuanta memoria adicional se utilizo en comparación al uso de memoria antes de ejecutar el codigo lo cual indica que no se necesitó mas de 1 MB adicional para ejecutar este proceso.

In [14]:
# q3t
pr.enable()
resultq3t = q3t.q3_time(file_path)
pr.disable()


s3 = io.StringIO()

# Crea un objeto Stats y usa print_stats para imprimir el informe
ps3 = pstats.Stats(pr, stream=s3).sort_stats('cumulative')
ps3.print_stats()


In [15]:
print(s3.getvalue())

         3060761 function calls (3059839 primitive calls) in 24.090 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        6    0.000    0.000   24.089    4.015 /home/josed/Desafio/venv/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3541(run_code)
        6    0.000    0.000   24.089    4.015 {built-in method builtins.exec}
        1    0.000    0.000   12.142   12.142 /tmp/ipykernel_152291/1963347963.py:1(<module>)
        1    0.000    0.000   12.142   12.142 /home/josed/Desafio/src/q1_time.py:8(q1_time)
        2    0.015    0.007    9.929    4.964 /home/josed/Desafio/venv/lib/python3.10/site-packages/pandas/io/json/_json.py:505(read_json)
      217    0.001    0.000    9.424    0.043 /usr/lib/python3.10/socket.py:691(readinto)
      217    9.423    0.043    9.423    0.043 {method 'recv_into' of '_socket.socket' objects}
      212    0.001    0.000    9.416    0.044 /home/josed/Desafio/venv/lib/python3.

In [16]:
resultq3t

[('@narendramodi', 2261),
 ('@Kisanektamorcha', 1836),
 ('@RakeshTikaitBKU', 1639),
 ('@PMOIndia', 1422),
 ('@RahulGandhi', 1125),
 ('@GretaThunberg', 1046),
 ('@RaviSinghKA', 1015),
 ('@rihanna', 972),
 ('@UNHumanRights', 962),
 ('@meenaharris', 925)]

In [17]:
%%memit
resultq3m = q3m.q3_memory(file_path)


peak memory: 1018.96 MiB, increment: 127.66 MiB


In [18]:
#q3m
resultq3m


[('@narendramodi', 2261),
 ('@Kisanektamorcha', 1836),
 ('@RakeshTikaitBKU', 1639),
 ('@PMOIndia', 1422),
 ('@RahulGandhi', 1125),
 ('@GretaThunberg', 1046),
 ('@RaviSinghKA', 1015),
 ('@rihanna', 972),
 ('@UNHumanRights', 962),
 ('@meenaharris', 925)]

### Opciones de mejora(punto 6.):

1) Escalabilidad de Recursos: La capacidad de procesamiento actual está limitada por las especificaciones del hardware en uso. Al migrar la ejecución de las funciones a un entorno en la nube, como AWS Glue, se podría aprovechar su arquitectura distribuida para escalar los recursos conforme sea necesario. Esto incluiría el aumento de CPU y memoria RAM disponibles, lo que potencialmente reduciría los tiempos de ejecución. Aunque no representa una optimización algorítmica per se, esta acción mejora el rendimiento al proporcionar acceso a una infraestructura más robusta.
2) Integración con Fuentes de Datos en Tiempo Real: En la situación actual, se manipulan datos estáticos extraídos de un único archivo. Si se obtuviera acceso a la API de Twitter, sería posible consumir datos en tiempo real y adaptar la recolección a necesidades específicas. Esto no solo proporcionaría información más actualizada, sino que también permitiría una mayor personalización en la captura y análisis de los datos.
3) Mejora en la Visualización y Reportes: Para obtener insights más profundos y facilitar la interpretación de los resultados, se podría desarrollar un sistema de reportes mejorado. Este sistema presentaría los datos de manera más detallada y accesible, utilizando visualizaciones interactivas que faciliten la comprensión de las métricas y tendencias clave. Tal sistema podría incorporar herramientas de BI para un análisis más sofisticado y apoyar la toma de decisiones basada en datos.
4) Estandarización del Código Mediante Bibliotecas únicas: Actualmente, es posible que el proyecto utilice múltiples bibliotecas para realizar tareas similares o relacionadas, lo cual puede generar redundancia y aumentar la complejidad del código. Para una proxima iteración, se recomienda evaluar y seleccionar una única biblioteca o conjunto de herramientas cohesivo que pueda satisfacer todos los requisitos funcionales. Esto facilitaría la mantenibilidad del código, mejoraría la claridad para nuevos desarrolladores que se incorporen al proyecto y optimizaría las dependencias, reduciendo así los potenciales conflictos y simplificando el proceso de actualización de las bibliotecas.